### PR TITLE
Adds AWS Budgets support to the configuration file

### PR DESCRIPTION
**Pull request by danielslbsb**
### Description of changes
* This adds the necessary budget fields schema and configuration classes, and creates a budget builder file, for the Cfn template.
* Added basic field validation to the schema and unit tests to ensure they are working correctly.

### Tests
* Manual testing (creating clusters and updating configuration after creation)
  * Creating cluster without budget and adding it after creation
  * Creating cluster with a budget, then changing budget category / properties
  * Creating a cluster with a budget, then deleting the budget
* Unit test to check basic validation of the budget schema

### References
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist
- [ ] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [ ] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 1]
```diff
@@ -925,13 +930,119 @@ def make_resource(self, data, **kwargs):
...
+    @post_load
+    def make_resource(self, data, **kwargs):
+        """Generate resource."""
+        return BudgetLimit(**data)
+
+
+class BudgetSchema(BaseSchema):
+    """Represent the schema of an aws budget."""
+
+    budget_category = fields.Str(
```
Comment (chenwany) [inline cli/src/pcluster/schemas/cluster_schema.py:1012]: We can add a validation in the schema level to validate budget_category = cluster and queue_name fields are not coexist, something like:
```
    @validates_schema
    def no_coexist_....
```
An example from validate the scheduler fields:
```
    @validates_schema
    def no_coexist_schedulers(self, data, **kwargs):
```

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 2]
```diff
@@ -925,13 +930,119 @@ def make_resource(self, data, **kwargs):
...
+        required=True,
+        metadata={"update_policy": UpdatePolicy.SUPPORTED},
+        validate=validate.OneOf(["cluster", "custom", "queue"]),
+    )
+    queue_name = fields.Str(metadata={"update_policy": UpdatePolicy.SUPPORTED})
+
+    cost_filters = fields.Dict(metadata={"update_policy": UpdatePolicy.SUPPORTED})
```
Comment (chenwany) [inline cli/src/pcluster/schemas/cluster_schema.py:1019]: cost filter fields can only be coexist with BudgetCategory: custom, let's add a schema validation

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 3]
```diff
@@ -925,13 +930,119 @@ def make_resource(self, data, **kwargs):
...
+        """Generate resource."""
+        return BudgetSubscriber(**data)
+
+
+class BudgetNotificationWithSubscribersSchema(BaseSchema):
+    """Represent the schema of the NotificationsWithSubscriber field of a budget."""
+
+    notification = fields.Nested(
+        BudgetNotificationSchema,
+        equired=True,
```
Comment (chenwany) [inline cli/src/pcluster/schemas/cluster_schema.py]: Typo `equired` here?

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 2 above)
Comment (chenwany) [inline cli/src/pcluster/schemas/cluster_schema.py:1019]: Let's add a schema validation for the supported keys in cost_filters":

Here's an example of validation metadata dict
```
    @validates("metadata")
    def validate_metadata(self, value):
        """Validate metadata contains fieds 'name' and 'version'."""
        for key in ["Name", "Version"]:
            if key not in value.keys():
                raise ValidationError(f"{key} is required for scheduler plugin Metadata.")
```

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 4]
```diff
@@ -925,13 +930,119 @@ def make_resource(self, data, **kwargs):
...
+    @post_load
+    def make_resource(self, data, **kwargs):
+        """Generate resource."""
+        return BudgetNotificationWithSubscribers(**data)
+
+
+class BudgetLimitSchema(BaseSchema):
+    """Represent the schema of the BudgetLimit field of a budget."""
+
+    amount = fields.Int(required=True, metadata={"update_policy": UpdatePolicy.SUPPORTED})
```
Comment (chenwany) [inline cli/src/pcluster/schemas/cluster_schema.py]: Is this amount require to be integer? What about float?

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 5]
```diff
@@ -925,13 +930,119 @@ def make_resource(self, data, **kwargs):
...
+    unit = fields.Str(metadata={"update_policy": UpdatePolicy.SUPPORTED})
```
Comment (chenwany) [inline cli/src/pcluster/schemas/cluster_schema.py]: Do we know how many Unit and what Unit are supported here?

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 6]
```diff
@@ -925,13 +930,119 @@ def make_resource(self, data, **kwargs):
...
+
+    notification_type = fields.Str(
+        metadata={"update_policy": UpdatePolicy.SUPPORTED},
+        validate=validate.OneOf(["ACTUAL", "FORECASTED"]),
+    )
+    comparison_operator = fields.Str(
+        metadata={"update_policy": UpdatePolicy.SUPPORTED},
+        validate=validate.OneOf(["GREATER_THAN", "EQUAL_TO", "LESS_THAN"]),
+    )
+    threshold = fields.Int(required=True, metadata={"update_policy": UpdatePolicy.SUPPORTED})
```
Comment (chenwany) [inline cli/src/pcluster/schemas/cluster_schema.py]: Does `threshold` require to be an interger? Can it be float?

File: cli/src/pcluster/config/cluster_config.py [hunk 7]
```diff
@@ -856,6 +856,71 @@ def __init__(self, head_node_bootstrap_timeout: int = None, compute_node_bootstr
...
 
 
+class BudgetNotification(Resource):
+    """Represent the configuration of each notification under the NotificationsWithSubscribers field of a budget."""
+
+    def __init__(
+        self,
+        notification_type: str = None,
+        comparison_operator: str = None,
+        threshold: int = None,
```
Comment (chenwany) [inline cli/src/pcluster/config/cluster_config.py]: Please double check the threshold type here

File: cli/src/pcluster/config/cluster_config.py [hunk 8]
```diff
@@ -856,6 +856,71 @@ def __init__(self, head_node_bootstrap_timeout: int = None, compute_node_bootstr
...
+    def __init__(self, notification: BudgetNotification = None, subscribers: List[BudgetSubscriber] = None):
+        super().__init__()
+        self.notification = notification
+        self.subscribers = subscribers
+
+
+class BudgetLimit(Resource):
+    """Represent the configuration of a budget limit."""
+
+    def __init__(self, amount: int = None, unit: str = None):
```
Comment (chenwany) [inline cli/src/pcluster/config/cluster_config.py]: Please double check the amount type here

File: cli/src/pcluster/templates/cluster_stack.py [hunk 9]
```diff
@@ -289,6 +290,11 @@ def _add_resources(self):
                 cw_log_group_name=self.log_group.log_group_name if self.config.is_cw_logging_enabled else None,
             )
 
+        # Budgets
+        if self.config.dev_settings and self.config.dev_settings.budgets:
+            cost_budgets = CostBudgets(self, self.config)
+            self.budget_list = cost_budgets.create_budgets()
```
Comment (chenwany) [inline cli/src/pcluster/templates/cluster_stack.py]: Shall we move the create_budgets function into budget_builder.py? 
Add the end of the _init_ function, we can add the resources, as what we do for other builder stack

File: cli/src/pcluster/templates/budget_builder.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        self.scope = scope
+
+    def create_budgets(self):
+        """Create the Cfn templates for the list of budgets."""
+        budget_list = self.cluster_config.dev_settings.budgets
+        cfn_budget_list = []
+        for index, budget in enumerate(budget_list):
+
+            budget_data = budgets.CfnBudget.BudgetDataProperty(
+                budget_name=f"CfnBudget{self.cluster_config.cluster_name}" + str(index),
```
Comment (chenwany) [inline cli/src/pcluster/templates/budget_builder.py]: If the budget name only contains cluster_name and index, we can not recognize whether the budget is queue budget, custom budget or cluster budget from its name. Let's make the budget_name more clear

File: cli/src/pcluster/templates/budget_builder.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+                                budgets.CfnBudget.SubscriberProperty(
+                                    address=subscriber.address,
+                                    subscription_type=subscriber.subscription_type,
+                                )
+                                for subscriber in single_notification.subscribers
+                            ],
+                        )
+                    )
+
+            budget_id = f"CfnBudget{self.cluster_config.cluster_name}" + str(index)
```
Comment (chenwany) [inline cli/src/pcluster/templates/budget_builder.py]: same comment for budget_id

File: cli/src/pcluster/templates/budget_builder.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+
+class CostBudgets:
+    """Create the budgets based on the configuration file."""
+
+    def __init__(self, scope: Construct, cluster_config: BaseClusterConfig):
+        self.cluster_config = cluster_config
+        self.scope = scope
+
+    def create_budgets(self):
```
Comment (chenwany) [inline cli/src/pcluster/templates/budget_builder.py]: The `create_budgets` function is long and contains many things, let's break it into several small functions.
First, let's rename it to _add_resources(), as we do for other builder stack files.
Then you can think about breaking the function into:
Add budget_data, add notifications_with_subscribers...
Or think about breaking it into: add_cluster_budget, add_queue_budget, add_custom_budget...

**Comment by chenwany** [comment]
Good job! Let's also add these fields into
example_configs awsbatch.full.yaml, slurm.full.yaml and scheduler_plugin.full.yaml

**Comment by chenwany** [comment]
In the PR description, could you list the manually test cases you perform?

**Comment by chenwany** [comment]
Action item: add validators for the fields and integration test in another PR

File: cli/src/pcluster/templates/budget_builder.py (hunk 12 above)
Comment (danielslbsb) [reply cli/src/pcluster/templates/budget_builder.py]: Good call! Restructured it.

File: cli/src/pcluster/templates/budget_builder.py (hunk 10 above)
Comment (danielslbsb) [reply cli/src/pcluster/templates/budget_builder.py]: Good idea. I made it more descriptive

File: cli/src/pcluster/templates/cluster_stack.py (hunk 9 above)
Comment (danielslbsb) [reply cli/src/pcluster/templates/cluster_stack.py]: Good call. Done!

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 5 above)
Comment (danielslbsb) [reply cli/src/pcluster/schemas/cluster_schema.py]: Not really, here's the documentation on that: https://docs.aws.amazon.com/aws-cost-management/latest/APIReference/API_budgets_Spend.html

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 1 above)
Comment (danielslbsb) [reply cli/src/pcluster/schemas/cluster_schema.py:1012]: Good idea. Done!

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 2 above)
Comment (danielslbsb) [reply cli/src/pcluster/schemas/cluster_schema.py:1019]: Good call. Added!

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 3 above)
Comment (danielslbsb) [reply cli/src/pcluster/schemas/cluster_schema.py]: Fixed!

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 2 above)
Comment (danielslbsb) [reply cli/src/pcluster/schemas/cluster_schema.py:1019]: Done!

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 4 above)
Comment (danielslbsb) [reply cli/src/pcluster/schemas/cluster_schema.py]: Good call, it's actually a float. Updated!

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 5 above)
Comment (danielslbsb) [reply cli/src/pcluster/schemas/cluster_schema.py]: From testing it, it seems that the only currently accepted one is USD, so I added a check for that.

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 6 above)
Comment (danielslbsb) [reply cli/src/pcluster/schemas/cluster_schema.py]: It should be a float. Updated!

File: cli/src/pcluster/config/cluster_config.py (hunk 7 above)
Comment (danielslbsb) [reply cli/src/pcluster/config/cluster_config.py]: Done!

File: cli/src/pcluster/config/cluster_config.py (hunk 8 above)
Comment (danielslbsb) [reply cli/src/pcluster/config/cluster_config.py]: Done!

File: cli/src/pcluster/templates/budget_builder.py (hunk 11 above)
Comment (danielslbsb) [reply cli/src/pcluster/templates/budget_builder.py]: Updated!

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 13]
```diff
@@ -925,13 +930,157 @@ def make_resource(self, data, **kwargs):
...
+
+    notification_type = fields.Str(
+        metadata={"update_policy": UpdatePolicy.SUPPORTED},
+        validate=validate.OneOf(["ACTUAL", "FORECASTED"]),
+    )
+    comparison_operator = fields.Str(
+        metadata={"update_policy": UpdatePolicy.SUPPORTED},
+        validate=validate.OneOf(["GREATER_THAN", "EQUAL_TO", "LESS_THAN"]),
+    )
+    threshold = fields.Float(required=True, metadata={"update_policy": UpdatePolicy.SUPPORTED})
```
Comment (chenwany) [inline cli/src/pcluster/schemas/cluster_schema.py]: nit: we can add `validate=validate.Range(min=0)`

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 14]
```diff
@@ -925,13 +930,157 @@ def make_resource(self, data, **kwargs):
...
+    def make_resource(self, data, **kwargs):
+        """Generate resource."""
+        return BudgetNotificationWithSubscribers(**data)
+
+
+class BudgetLimitSchema(BaseSchema):
+    """Represent the schema of the BudgetLimit field of a budget."""
+
+    amount = fields.Float(required=True, metadata={"update_policy": UpdatePolicy.SUPPORTED})
+    unit = fields.Str(metadata={"update_policy": UpdatePolicy.SUPPORTED}, validate=validate.OneOf(["USD"]))
```
Comment (chenwany) [inline cli/src/pcluster/schemas/cluster_schema.py]: Do we want to restrict unit to only be USD?

File: cli/src/pcluster/models/cluster.py [hunk 15]
```diff
@@ -435,7 +435,7 @@ def _validate_and_parse_config(self, validator_suppressors, validation_failure_l
 
     @staticmethod
     def _load_additional_instance_type_data(cluster_config_dict):
-        if "DevSettings" in cluster_config_dict:
```
Comment (chenwany) [inline cli/src/pcluster/models/cluster.py:438]: Let change it to something like:
```
        if "DevSettings" in cluster_config_dict and cluster_config_dict.get("DevSettings"):
            instance_types_data = cluster_config_dict.get("DevSettings").get("InstanceTypesData")
```

File: cli/src/pcluster/templates/cluster_stack.py [hunk 16]
```diff
@@ -289,6 +290,10 @@ def _add_resources(self):
                 cw_log_group_name=self.log_group.log_group_name if self.config.is_cw_logging_enabled else None,
             )
 
+        # Budgets
+        if self.config.dev_settings and self.config.dev_settings.budgets:
+            CostBudgets(self, self.config)
```
Comment (chenwany) [inline cli/src/pcluster/templates/cluster_stack.py]: self.budgets=CostBudgets(self, self.config)

File: cli/src/pcluster/templates/budget_builder.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+        for index, budget in enumerate(budget_list):
+            cfn_budget = self._add_parameters(budget, index)
+            cfn_budget_list.append(cfn_budget)
+
+        return cfn_budget_list
+
+    def _add_parameters(self, budget, index):
+        """Add each budget's parameters."""
+        budget_data = budgets.CfnBudget.BudgetDataProperty(
+            budget_name=f"CfnBudget{str(index)}{self.cluster_config.cluster_name}{budget.budget_category}",
```
Comment (chenwany) [inline cli/src/pcluster/templates/budget_builder.py]: Let's remove CfnBudget from the name , the name can be {cluster-name}-{budget-category}-{region}

File: cli/src/pcluster/templates/budget_builder.py [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+        )
+        notifications_with_subscribers = add_budget_notifications(budget)
+        budget_id = f"CfnBudget{str(index)}{self.cluster_config.cluster_name}{budget.budget_category}"
+        cfn_budget = budgets.CfnBudget(
+            self.scope, budget_id, budget=budget_data, notifications_with_subscribers=notifications_with_subscribers
+        )
+        return cfn_budget
+
+
+def add_budget_notifications(budget):
```
Comment (chenwany) [inline cli/src/pcluster/templates/budget_builder.py:62]: This function is returning either None or a list. It is in general better to have the function return single type of data.
Instead you can move the if condition out of the function:
```
notifications_with_subscribers = add_budget_notifications(budget) if budget.notifications_with_subscribers is not None else None
```
Then in the function you ca remove if

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 14 above)
Comment (danielslbsb) [reply cli/src/pcluster/schemas/cluster_schema.py]: From testing it, it seems that it is the only currently accepted value.

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 13 above)
Comment (danielslbsb) [reply cli/src/pcluster/schemas/cluster_schema.py]: Added!

File: cli/src/pcluster/models/cluster.py (hunk 15 above)
Comment (danielslbsb) [reply cli/src/pcluster/models/cluster.py:438]: Updated! Simplified it a bit.

File: cli/src/pcluster/templates/cluster_stack.py (hunk 16 above)
Comment (danielslbsb) [reply cli/src/pcluster/templates/cluster_stack.py]: Fixed!

File: cli/src/pcluster/templates/budget_builder.py (hunk 17 above)
Comment (danielslbsb) [reply cli/src/pcluster/templates/budget_builder.py]: Improved naming!

File: cli/src/pcluster/templates/budget_builder.py (hunk 18 above)
Comment (danielslbsb) [reply cli/src/pcluster/templates/budget_builder.py:62]: Updated!

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 19]
```diff
@@ -925,13 +930,161 @@ def make_resource(self, data, **kwargs):
...
+        """Generate resource."""
+        return Budget(**data)
+
+    @validates_schema
+    def no_category_conflict(self, data, **kwargs):
+        """Validate that only custom budgets have CostFilters, and only queue budgets have QueueName."""
+        budget_category = data.get("budget_category")
+        cost_filters = data.get("cost_filters")
+        queue_name = data.get("queue_name")
+        if budget_category != "custom" and cost_filters is not None:
```
Comment (chenwany) [inline cli/src/pcluster/schemas/cluster_schema.py]: can we remove is not None here?
Can we use if else here?
For example:
```
if budget_category != "custom" and cost_filters:
    raise ValidationError("CostFilters can not be configured when BudgetCategory is set to Custom.")
elif budget_category == "custom" and not cost_filters:
    raise ValidationError("CostFilters is required when BudgetCategory is set to Custom.")
if budget_category != "queue" and not queue_name:
    raise ValidationError("QueueName can only be specified when BudgetCategory is set to Queue.")
elif budget_category == "queue" and queue_name is None:
    raise ValidationError("Queue budgets require QueueName is required when BudgetCategory is set to Queue.")
```

File: cli/src/pcluster/templates/budget_builder.py [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+            cfn_budget_list.append(cfn_budget)
+
+        return cfn_budget_list
+
+    def _add_parameters(self, budget, index):
+        """Add each budget's parameters."""
+        if budget.budget_category == "queue":
+            name = f"{self.cluster_config.cluster_name}-{index}-{budget.queue_name} {self.cluster_config.region}"
+        elif budget.budget_category == "cluster":
+            name = f"{self.cluster_config.cluster_name}-{index} {self.cluster_config.region}"
```
Comment (chenwany) [inline cli/src/pcluster/templates/budget_builder.py]: Missing `-` between {index} and region?

File: cli/src/pcluster/templates/budget_builder.py [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+        else:
+            name = f"{self.cluster_config.cluster_name}-{index}-custom {self.cluster_config.region}"
```
Comment (chenwany) [inline cli/src/pcluster/templates/budget_builder.py]: Add `-` between custom and region?

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 19 above)
Comment (danielslbsb) [reply cli/src/pcluster/schemas/cluster_schema.py]: Good suggestion. Updated!

File: cli/src/pcluster/templates/budget_builder.py (hunk 20 above)
Comment (danielslbsb) [reply cli/src/pcluster/templates/budget_builder.py]: Updated!

File: cli/src/pcluster/templates/budget_builder.py (hunk 21 above)
Comment (danielslbsb) [reply cli/src/pcluster/templates/budget_builder.py]: Updated!

**Review by chenwany — APPROVED** [review]
Well done!

**Pull request merged**